File: js/CjkEncoder.js
goog.provide('twim.CjkEncoder');

goog.require('twim.CjkTransform');

/**
 * @param{!Uint8Array} data bytes
 * @return{!Array<number>} CJK code-points
 */
twim.CjkEncoder.encode = function(data) {
  let /** @type{number} */ dataBits = data.length * 8;
  let /** @type{!Uint16Array} */ nibbles = new Uint16Array(5 * Math.ceil(dataBits / 41) + 1);
  let /** @type{number} */ offset = 0;
  let /** @type{number} */ nibblesLength = 0;
  while (offset < dataBits) {
    let /** @type{number} */ p = 0;
    let /** @type{number} */ ix = offset >> 3;
    for (let /** @type{number} */ i = ix + 5; i >= ix; --i) {
      p = (p * 256) + ((i < data.length) ? data[i] : 0);
    }
    p = Math.floor(p / (1 << (offset & 7))) % 2199023255552;
    offset += 41;
    for (let /** @type{number} */ i = 0; i < 5; ++i) {
      nibbles[nibblesLength++] = p % 296;
      p = Math.floor(p / 296);
    }
  }
  let /** @type{!Array<number>} */ result = [];
  for (let /** @type{number} */ i = 0; i < nibblesLength; i += 2) {
    let /** @type{number} */ ord = nibbles[i] + 296 * nibbles[i + 1];
    result.push(twim.CjkTransform.ordinalToUnicode(ord));
  }
  let /** @type{number} */ unicodeZero = twim.CjkTransform.ordinalToUnicode(0);
  while (result[result.length - 1] === unicodeZero) {
    result.pop();
  }
  return result;
};

File: java/ru/eustas/twim/CjkEncoder.java
package ru.eustas.twim;

/**
 * Byte array to CJK string encoder.
 */
final class CjkEncoder {
  // Collection of utilities. Should not be instantiated.
  CjkEncoder() {}

  static int[] encode(byte[] data) {
    int bufferLength = (5 * (((data.length * 8) + 40) / 41) + 1) / 2;
    int[] nibbles = new int[2 * bufferLength];
    int nibbleIdx = 0;
    int offset = 0;
    while (offset < data.length * 8) {
      long p = 0;
      int ix = offset >> 3;
      for (int i = 0; i < 6; ++i) {
        int idx = ix + 5 - i;
        int b = (idx < data.length) ? (data[idx] & 0xFF) : 0;
        p = (p << 8) | b;
      }
      p = (p >> (offset & 7)) & 0x1FFFFFFFFFFL;
      offset += 41;
      for (int i = 0; i < 5; ++i) {
        nibbles[nibbleIdx++] = (int) (p % 296L);
        p = p / 296L;
      }
    }
    int[] buffer = new int[bufferLength];
    for (int i = 0; i < nibbles.length; i += 2) {
      buffer[i / 2] = CjkTransform.ordinalToUnicode(nibbles[i] + 296 * nibbles[i + 1]);
    }
    int unicodeZero = CjkTransform.ordinalToUnicode(0);
    while (bufferLength > 0 && buffer[bufferLength - 1] == unicodeZero) {
      bufferLength--;
    }
    int[] result = new int[bufferLength];
    System.arraycopy(buffer, 0, result, 0, bufferLength);
    return result;
  }
}